production/ksonnet/loki/boltdb_shipper.libsonnet

```diff
@@ -0,0 +1,143 @@
+    pvc.mixin.spec.resources.withRequests({ storage: $._config.ingester_pvc_size }) +
+    pvc.mixin.spec.withAccessModes(['ReadWriteOnce']) +
+    pvc.mixin.spec.withStorageClassName('fast'),
+
+  ingester_deployment: {},
+
+  ingester_args+:: {
+    // Persist index in pvc
+    'boltdb.shipper.active-index-directory': '/data/index',
+
+    // Use PVC for caching
+    'boltdb.shipper.cache-location': '/data/boltdb-cache',
+
+    // Disable chunk transfer
+    'ingester.max-transfer-retries': 0,
+  },
+
+  ingester_container+::
+    container.withVolumeMountsMixin([
+      volumeMount.new('ingester-data', '/data'),
+    ]),
+
+  ingester_statefulset:
+    statefulSet.new('ingester', 3, [$.ingester_container], ingester_data_pvc) +
+    statefulSet.mixin.spec.withServiceName('ingester') +
+    $.config_hash_mixin +
+    $.util.configVolumeMount('loki', '/etc/loki/config') +
+    $.util.configVolumeMount('overrides', '/etc/loki/overrides') +
+    $.util.antiAffinity +
+    statefulSet.mixin.spec.updateStrategy.withType('RollingUpdate') +
+    statefulSet.mixin.spec.template.spec.securityContext.withFsGroup(10001) +  // 10001 is the group ID assigned to Loki in the Dockerfile
+    statefulSet.mixin.spec.template.spec.withTerminationGracePeriodSeconds(4800),
+
+  ingester_service:
+    $.util.serviceFor($.ingester_statefulset),
+
+  // Use PVC for queriers instead of node disk.
+  local querier_data_pvc =
+    pvc.new('querier-data') +
+    pvc.mixin.spec.resources.withRequests({ storage: $._config.querier_pvc_size }) +
+    pvc.mixin.spec.withAccessModes(['ReadWriteOnce']) +
+    pvc.mixin.spec.withStorageClassName('fast'),
+
+  querier_deployment: {},
+
+  querier_args+:: {
+    // Use PVC for caching
+    'boltdb.shipper.cache-location': '/data/boltdb-cache',
+  },
+
+  querier_container+::
+    container.withVolumeMountsMixin([
+      volumeMount.new('querier-data', '/data'),
+    ]),
+
+  querier_statefulset:
+    statefulSet.new('querier', 3, [$.querier_container], querier_data_pvc) +
+    statefulSet.mixin.spec.withServiceName('querier') +
+    $.config_hash_mixin +
+    $.util.configVolumeMount('loki', '/etc/loki/config') +
+    $.util.configVolumeMount('overrides', '/etc/loki/overrides') +
+    $.util.antiAffinity +
+    statefulSet.mixin.spec.updateStrategy.withType('RollingUpdate') +
+    statefulSet.mixin.spec.template.spec.securityContext.withFsGroup(10001),  // 10001 is the group ID assigned to Loki in the Dockerfile
+
+  querier_service:
+    $.util.serviceFor($.querier_statefulset),
+
+  memcached_index_writes:: {}, // we don't dedupe index writes when using boltdb-shipper so don't deploy a cache for it.
+
+  // Use PVC for compactor instead of node disk.
+  local compactor_data_pvc =
+    pvc.new('compactor-data') +
+    pvc.mixin.spec.resources.withRequests({ storage: $._config.compactor_pvc_size }) +
+    pvc.mixin.spec.withAccessModes(['ReadWriteOnce']) +
+    pvc.mixin.spec.withStorageClassName('fast'),
+
+  compactor_args::
```

Comment:
Again, I think we should prefer config file > args

Reply:
I have kept service-specific configs as args to avoid causing all the services to be rolled out un-necessarily due to config change and it lets us keep service-specific args not to be passed to every other service. I don't mind changing it but it would good to hear what do you think about it?